pageserver/src/relish_storage.rs

```diff
@@ -0,0 +1,103 @@
+//! Abstractions for the page server to store its relish layer data in the external storage.
+//!
+//! Main purpose of this module subtree is to provide a set of abstractions to manage the storage state
+//! in a way, optimal for page server.
+//!
+//! The abstractions hide multiple custom external storage API implementations,
+//! such as AWS S3, local filesystem, etc., located in the submodules.
+
+mod local_fs;
+mod rust_s3;
+/// A queue and the background machinery behind it to upload
+/// local page server layer files to external storage.
+pub mod storage_uploader;
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+/// Storage (potentially remote) API to manage its state.
+#[async_trait::async_trait]
+pub trait RelishStorage: Send + Sync {
+    async fn list_relishes(&self) -> anyhow::Result<Vec<Arc<dyn RelishStoragePath>>>;
+
+    async fn download_relish(&self, from: &dyn RelishStoragePath, to: &Path) -> anyhow::Result<()>;
+
+    async fn delete_relish(&self, path: &dyn RelishStoragePath) -> anyhow::Result<()>;
+
+    async fn upload_relish(
+        &self,
+        relish_local_path: &Path,
+        destination: StorageDestination,
+    ) -> anyhow::Result<()>;
+}
+
+/// A unique path of the relish data in the storage.
+pub trait RelishStoragePath: Send + Sync {
+    /// Unique key in the storage for the relish data.
+    fn key(&self) -> &str;
+}
+
+/// A wrapper around [`PathBuf`] used by storage to determine where to place the relish data.
+/// The wrapper ensures a number of invariants:
+///
+/// * path is relative
+/// * path points to a file, that has a non-empty name
+#[derive(Debug)]
+pub struct StorageDestination {
```

Comment:
It feels like `RelishStoragePath` and `StorageDestination` are representing the same thing. Would it make more sense if `RelishStorage` had a symmetric API?
```
fn download(from: RemoteKey, to: LocalPath)
fn upload(from: LocalPath, to: RemoteKey)
```